File: backend/verification-worker/src/sui/suiProofClient.ts
import { SuiClient } from "@mysten/sui/client";
import { Transaction } from "@mysten/sui/transactions";
import { Ed25519Keypair } from "@mysten/sui/keypairs/ed25519";
import { decodeSuiPrivateKey } from "@mysten/sui/cryptography";

import type { Logger } from "../logging/logger";

export type ProofCreationResult = {
  txDigest: string;
  objectId: string;
};

function hexToBytes32(hex: string): Uint8Array {
  const h = String(hex || "").trim().toLowerCase();
  if (!/^[0-9a-f]{64}$/.test(h)) throw new Error("Invalid data_hash hex (expected 64 hex chars)");
  return Uint8Array.from(Buffer.from(h, "hex"));
}

function parseEd25519Keypair(privateKey: string): Ed25519Keypair {
  const raw = String(privateKey || "").trim();
  if (!raw) throw new Error("Missing SUI_PRIVATE_KEY");

  // Prefer Sui CLI exported format: "suiprivkey..."
  if (raw.startsWith("suiprivkey")) {
    const decoded = decodeSuiPrivateKey(raw);
    if (decoded.schema !== "ED25519") throw new Error(`Unsupported key scheme: ${decoded.schema}`);
    return Ed25519Keypair.fromSecretKey(decoded.secretKey);
  }

  // Fallback: base64-encoded raw secret key bytes.
  return Ed25519Keypair.fromSecretKey(Uint8Array.from(Buffer.from(raw, "base64")));
}

function extractCreatedObjectId(objectChanges: any[] | undefined, expectedSuffix: string): string {
  const changes = Array.isArray(objectChanges) ? objectChanges : [];
  for (const c of changes) {
    if (!c || c.type !== "created") continue;
    const objectType = String(c.objectType || "");
    if (!objectType) continue;
    if (!objectType.endsWith(expectedSuffix)) continue;
    const objectId = String(c.objectId || "");
    if (objectId) return objectId;
  }
  throw new Error("Created proof object id not found in transaction object changes");
}

export class SuiProofClient {
  private readonly client: SuiClient;
  private readonly signer: Ed25519Keypair;

  constructor(
    private readonly logger: Logger,
    opts: {
      rpcUrl: string;
      privateKey: string;
      packageId: string;
      moduleName: string;
      functionName: string;
    }
  ) {
    this.client = new SuiClient({ url: opts.rpcUrl });
    this.signer = parseEd25519Keypair(opts.privateKey);
    this._packageId = opts.packageId;
    this._moduleName = opts.moduleName;
    this._functionName = opts.functionName;
  }

  private readonly _packageId: string;
  private readonly _moduleName: string;
  private readonly _functionName: string;

  async createProof(dataHashHex: string, createdAtSeconds: number): Promise<ProofCreationResult> {
    const hashBytes = hexToBytes32(dataHashHex);
    const createdAt = BigInt(Math.max(0, Math.trunc(createdAtSeconds || 0)));

    const tx = new Transaction();
    // Keep gas usage predictable; override via env if needed.
    tx.setGasBudget(BigInt(Math.max(1, Math.trunc(Number(process.env.SUI_GAS_BUDGET || 50_000_000)))));
    tx.moveCall({
      target: `${this._packageId}::${this._moduleName}::${this._functionName}`,
      arguments: [tx.pure.vector("u8", Array.from(hashBytes)), tx.pure.u64(createdAt)],
    });

    const res: any = await this.client.signAndExecuteTransaction({
      signer: this.signer,
      transaction: tx,
      options: {
        showEffects: true,
        showObjectChanges: true,
      },
    });

    const txDigest = String(res?.digest || "").trim();
    if (!txDigest) throw new Error("Missing transaction digest");

    // Proof type suffix: ::parlay_proof::Proof (package address varies)
    const objectId = extractCreatedObjectId(res?.objectChanges, `::${this._moduleName}::Proof`);

    this.logger.info({ txDigest, objectId }, "verification proof created");
    return { txDigest, objectId };
  }
}

File: backend/app/services/verification_records/sui_proof_client.py
"""Thin wrapper around pysui for SUI create_proof (parlay verification).

Matches the contract of the TS verification worker: create_proof(data_hash_hex, created_at_seconds)
returns (tx_digest, object_id). Used by the Pattern A verifier (sui_verifier.py).
"""

from __future__ import annotations

import logging
import os
from dataclasses import dataclass
from typing import Optional

logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class ProofCreationResult:
    tx_digest: str
    object_id: str


def _hex_to_bytes32(hex_str: str) -> bytes:
    h = (hex_str or "").strip().lower()
    if len(h) != 64 or not all(c in "0123456789abcdef" for c in h):
        raise ValueError("Invalid data_hash hex (expected 64 hex chars)")
    return bytes.fromhex(h)


def _extract_created_object_id(object_changes: list, expected_suffix: str) -> str:
    for c in object_changes or []:
        if not c or c.get("type") != "created":
            continue
        object_type = (c.get("objectType") or "").strip()
        if not object_type.endswith(expected_suffix):
            continue
        object_id = (c.get("objectId") or "").strip()
        if object_id:
            return object_id
    raise ValueError("Created proof object id not found in transaction object changes")


def _normalize_private_key_for_pysui(raw: str) -> str:
    """Return a key format pysui accepts: base64 keystring (key_type_flag | private_key_seed)."""
    raw = (raw or "").strip()
    if not raw:
        raise ValueError("Missing SUI_PRIVATE_KEY")
    if raw.startswith("suiprivkey"):
        try:
            import base64
            from pysui.sui.sui_crypto import KeyPair
            decoded = KeyPair.decode_sui_private_key(raw)
            if hasattr(decoded, "secret_key"):
                return base64.b64encode(getattr(decoded, "secret_key", decoded)).decode()
            if isinstance(decoded, bytes):
                return base64.b64encode(decoded).decode()
        except Exception:
            pass
    return raw


def create_proof(
    data_hash_hex: str,
    created_at_seconds: int,
    *,
    rpc_url: str,
    private_key: str,
    package_id: str,
    module_name: str,
    function_name: str,
    gas_budget: Optional[int] = None,
) -> ProofCreationResult:
    """Create an on-chain proof for a verification record.

    Calls package_id::module_name::function_name(data_hash: vector<u8>, created_at: u64).
    Returns tx_digest and the created Proof object id.
    """
    try:
        from pysui import SuiClient, SuiConfig, handle_result
        from pysui.sui.sui_txn import SyncTransaction
        from pysui.sui.sui_types import SuiU64
    except ImportError as e:
        raise RuntimeError("pysui is required for SUI verification; install with: pip install pysui") from e

    hash_bytes = _hex_to_bytes32(data_hash_hex)
    created_at = max(0, int(created_at_seconds or 0))
    budget = gas_budget or int(os.environ.get("SUI_GAS_BUDGET", "50000000"))

    key_for_config = _normalize_private_key_for_pysui(private_key)
    cfg = SuiConfig.user_config(rpc_url=rpc_url, prv_keys=[key_for_config])
    client = SuiClient(cfg)

    txn = SyncTransaction(client=client)
    target = f"{package_id}::{module_name}::{function_name}"
    txn.move_call(
        target=target,
        arguments=[list(hash_bytes), SuiU64(created_at)],
    )
    options = {"showEffects": True, "showObjectChanges": True}
    result = handle_result(txn.execute(gas_budget=str(budget), options=options))

    # handle_result may return SuiRpcResult; unwrap result_data if present.
    data = getattr(result, "result_data", result)
    j = (getattr(data, "to_json_dict", lambda: None)() or {}) if hasattr(data, "to_json_dict") else {}
    if not j and hasattr(result, "to_json_dict"):
        j = result.to_json_dict() or {}
    tx_digest = (j.get("digest") or getattr(data, "digest", "") or getattr(result, "digest", "") or "").strip()
    if not tx_digest:
        raise ValueError("Missing transaction digest in result")

    object_changes = j.get("objectChanges") or getattr(data, "object_changes", None) or getattr(result, "object_changes", None) or []
    proof_suffix = f"::{module_name}::Proof"
    object_id = _extract_created_object_id(list(object_changes), proof_suffix)

    logger.info("verification proof created tx_digest=%s object_id=%s", tx_digest, object_id)
    return ProofCreationResult(tx_digest=tx_digest, object_id=object_id)
